eZ/Publish/Core/Persistence/Legacy/Content/Gateway/DoctrineDatabase.php

```diff
@@ -346,249 +267,246 @@ public function updateContent($contentId, MetadataUpdateStruct $struct, VersionI
+            ->setParameter('content_id', $contentId, ParameterType::INTEGER)
+            ->setParameter('version_no', $versionNo, ParameterType::INTEGER);
+
+        $query->execute();
     }
 
-    /**
-     * Updates "always available" flag for Content identified by $contentId, in respect to
-     * Content's current main language and optionally new $alwaysAvailable state.
-     *
-     * @param int $contentId
-     * @param bool|null $alwaysAvailable New "always available" value or null if not defined
-     */
-    public function updateAlwaysAvailableFlag($contentId, $alwaysAvailable = null)
+    public function updateAlwaysAvailableFlag(int $contentId, ?bool $alwaysAvailable = null): void
     {
         // We will need to know some info on the current language mask to update the flag
         // everywhere needed
         $contentInfoRow = $this->loadContentInfo($contentId);
+        $versionNo = (int)$contentInfoRow['current_version'];
+        $languageMask = (int)$contentInfoRow['language_mask'];
+        $initialLanguageId = (int)$contentInfoRow['initial_language_id'];
         if (!isset($alwaysAvailable)) {
-            $alwaysAvailable = 1 === ($contentInfoRow['language_mask'] & 1);
+            $alwaysAvailable = 1 === ($languageMask & 1);
         }
 
-        /** @var $q \eZ\Publish\Core\Persistence\Database\UpdateQuery */
-        $q = $this->dbHandler->createUpdateQuery();
-        $q
-            ->update($this->dbHandler->quoteTable('ezcontentobject'))
+        $this->updateContentItemAlwaysAvailableFlag($contentId, $alwaysAvailable);
+        $this->updateContentNameAlwaysAvailableFlag(
+            $contentId,
+            $versionNo,
+            $alwaysAvailable
+        );
+        $this->updateContentFieldsAlwaysAvailableFlag(
+            $contentId,
+            $versionNo,
+            $alwaysAvailable,
+            $languageMask,
+            $initialLanguageId
+        );
+    }
+
+    private function updateContentItemAlwaysAvailableFlag(
+        int $contentId,
+        bool $alwaysAvailable
+    ): void {
+        $query = $this->connection->createQueryBuilder();
+        $expr = $query->expr();
+        $query
+            ->update(self::CONTENT_ITEM_TABLE)
             ->set(
-                $this->dbHandler->quoteColumn('language_mask'),
-                $alwaysAvailable ?
-                    $q->expr->bitOr($this->dbHandler->quoteColumn('language_mask'), 1) :
-                    $q->expr->bitAnd($this->dbHandler->quoteColumn('language_mask'), -2)
+                'language_mask',
+                $alwaysAvailable
+                    ? $this->databasePlatform->getBitOrComparisonExpression(
+                    'language_mask',
+                    ':languageMaskOperand'
+                )
+                    : $this->databasePlatform->getBitAndComparisonExpression(
+                    'language_mask',
+                    ':languageMaskOperand'
+                )
```

Comment:
Please move this expression outside of `set` call. In current form is unreadable.  _Comment is also applicable in other places as well._

Reply:
Refactored via 51d6370.